### PR TITLE
[CPU] Add support for Qwen3-vl and Qwen3-omni

### PR DESCRIPTION
**Pull request by blzheng**
## Motivation

This PR added frontend support for Qwen3-VL and Qwen3-Omni on CPU.

- Fixed the tp=3/6 padding issue in the vision encoder.
- Replaced nn.Linear with ReplicatedLinear 
- Applied optimization for image preprocessor, depends on #15168 
- Applied flash_attn_varlen_func for vision attention
- Applied linear_gelu_linear fusion for TP=1, depends on #17919 
- Applied optimization for conv3d, depends on #16040 
- Removed contiguous overhead before rope, depends on #13121 
- Applied optimization for layernorm, depends on #15075 

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #26430223189](https://github.com/sgl-project/sglang/actions/runs/26430223189)
Latest PR Test (Extra): :x: [Run #26430223129](https://github.com/sgl-project/sglang/actions/runs/26430223129)

### CODE REVIEW

**Review by mingfeima — APPROVED** [review]

**Comment by mingfeima** [comment]
@jianan-gu this pr adds `VisionAMXAttention`, is this a duplicate? i remembered that you have a similar one.

**Comment by jianan-gu** [comment]
> @jianan-gu this pr adds `VisionAMXAttention`, is this a duplicate? i remembered that you have a similar one.

no, it is not duplicate, previous one is just sdpa vision attention, we can switch to `VisionAMXAttention` from this PR.

**Comment by mingfeima** [comment]
@blzheng the linear-gelu-linear fusion part kind of like a hack right now, let's remove this part from the upstream PRs.

**Comment by blzheng** [comment]
> @blzheng the linear-gelu-linear fusion part kind of like a hack right now, let's remove this part from the upstream PRs.

Done.

File: python/sglang/srt/models/qwen3_omni_moe.py [hunk 1]
```diff
@@ -70,18 +85,37 @@ def __init__(
...
+                bias=True,
+                prefix=f"{prefix}.fc1",
+            )
+            self.fc2 = RowParallelLinear(
+                config.encoder_ffn_dim,
+                self.embed_dim,
+                quant_config=quant_config,
+                bias=True,
+                prefix=f"{prefix}.fc2",
+            )
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_omni_moe.py]:
```suggestion
        tp_size = get_tensor_model_parallel_world_size()
        use_replicated = config.encoder_ffn_dim % tp_size != 0
        fc1_cls = ReplicatedLinear if use_replicated else ColumnParallelLinear
        fc2_cls = ReplicatedLinear if use_replicated else RowParallelLinear
        self.fc1 = fc1_cls(
            self.embed_dim,
            config.encoder_ffn_dim,
            quant_config=quant_config,
            bias=True,
            prefix=f"{prefix}.fc1",
        )
        self.fc2 = fc2_cls(
            config.encoder_ffn_dim,
            self.embed_dim,
            quant_config=quant_config,
            bias=True,
            prefix=f"{prefix}.fc2",
        )
```

File: python/sglang/srt/multimodal/processors/qwen_vl.py [hunk 2]
```diff
@@ -59,6 +60,65 @@
...
+_is_cpu = is_cpu()
+if _is_cpu and _is_cpu_amx_available:
+    try:
+        import transformers
+        from transformers.image_processing_base import BatchFeature
+
+        image_preprocess_cpu = torch.ops.sgl_kernel.image_preprocess_cpu
+
+        def hacked_preprocess(
+            self,
```
Comment (mingfeima) [inline python/sglang/srt/multimodal/processors/qwen_vl.py]:
```suggestion
        def fast_preprocess_cpu(
            self,
```

File: python/sglang/srt/models/qwen3_omni_moe.py [hunk 3]
```diff
@@ -200,10 +234,19 @@ def __init__(self, config: Qwen3OmniMoeAudioEncoderConfig):
...
-        self.proj1 = nn.Linear(config.d_model, config.d_model)
+        self.conv_out = ReplicatedLinear(
+            conv_out_dim,
+            config.d_model,
+            bias=False,
+            quant_config=quant_config,
+        )
+        self.proj1 = ReplicatedLinear(
+            config.d_model, config.d_model, quant_config=quant_config
+        )
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_omni_moe.py:251]: why we need to change `ReplicatedLinear` here?
is it because nn.Linear won't go `weight_packed_linear`?

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/qwen3_omni_moe.py (hunk 3 above)
Comment (blzheng) [reply python/sglang/srt/models/qwen3_omni_moe.py:251]: Yes

File: python/sglang/srt/models/qwen3_vl.py [hunk 4]
```diff
@@ -97,6 +97,10 @@
 _is_cpu = is_cpu()
 
 
+_is_cpu = is_cpu()
+_is_cpu_amx_available = cpu_has_amx_support()
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_vl.py]:
```suggestion
_is_cpu_amx_available = cpu_has_amx_support()
_is_cpu = is_cpu()

```

Comment (mingfeima) [reply python/sglang/srt/models/qwen3_vl.py]: duplicated with L96 and L97

File: python/sglang/srt/multimodal/processors/qwen_vl.py [hunk 5]
```diff
@@ -59,6 +60,65 @@
...
+                patch_size,
+                temporal_patch_size,
+                merge_size,
+                True,
+                torch.bfloat16,
+            )
+            return BatchFeature(
+                data={"pixel_values": pixel_values, "image_grid_thw": image_grid_thw},
+                tensor_type=return_tensors,
+            )
```
Comment (mingfeima) [inline python/sglang/srt/multimodal/processors/qwen_vl.py]: we'd better put amx python wrapper in `python/sglang/srt/layers/amx_utils.py` and import from there.

File: python/sglang/srt/models/qwen3_omni_moe.py [hunk 6]
```diff
@@ -56,10 +65,16 @@ def __init__(
         super().__init__()
         embed_dim = config.d_model
         self.embed_dim = config.d_model
+        head_size = None
+        projection_size = embed_dim
+        if _is_cpu and hasattr(config, "original_encoder_attention_heads"):
+            head_size = embed_dim // config.original_encoder_attention_heads
+            projection_size = config.encoder_attention_heads * head_size
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_omni_moe.py]: it would be cleaner to isolate head size adjustment to an util before ``

File: python/sglang/srt/models/qwen3_omni_moe.py [hunk 7]
```diff
@@ -43,7 +50,9 @@
     Qwen3VLMoeForConditionalGeneration,
     load_fused_expert_weights,
 )
-from sglang.srt.utils import add_prefix, is_npu, logger
+from sglang.srt.utils import add_prefix, is_cpu, is_npu, logger
+
+_is_cpu = is_cpu()
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_omni_moe.py:56]:
```python
def get_head_dim_and_projection_size(
    embed_dim: int,
    num_heads: int,
    original_num_heads: Optional[int] = None,
) -> Tuple[Optional[int], int]:
    if (not _is_cpu) or original_num_heads is None:
        return None, embed_dim

    # On CPU, TP may pad num_heads (e.g. for tp=3/6). In that case we keep the
    # original per-head width (from original_num_heads) and recompute projection_size
    # with padded num_heads, so attention tensor shapes stay TP-friendly while
    # preserving checkpoint semantics.
    head_dim = embed_dim // original_num_heads
    projection_size = num_heads * head_dim
    return head_dim, projection_size
```


and call it:
```python
head_dim, projection_size = get_head_dim_and_projection_size(
    embed_dim=embed_dim,
    num_heads=config.encoder_attention_heads,
    original_num_heads=getattr(config, "original_encoder_attention_heads", None),
)
self.self_attn = VisionAttention(
    embed_dim=embed_dim,
    num_heads=config.encoder_attention_heads,
    head_dim=head_dim,
    projection_size=projection_size,
    use_qkv_parallel=True,
    proj_bias=True,
    flatten_batch=True,
    prefix=f"{prefix}.self_attn",
    quant_config=quant_config,
)
```

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/qwen3_omni_moe.py [hunk 8]
```diff
@@ -70,15 +104,21 @@ def __init__(
         self.dropout = config.dropout
         self.activation_fn = ACT2FN[config.activation_function]
         self.activation_dropout = config.activation_dropout
-        self.fc1 = ColumnParallelLinear(
+        tp_size = get_tensor_model_parallel_world_size()
+        use_replicated = config.encoder_ffn_dim % tp_size != 0
+        fc1_cls = ReplicatedLinear if use_replicated else ColumnParallelLinear
+        fc2_cls = ReplicatedLinear if use_replicated else RowParallelLinear
+        self.fc1 = fc1_cls(
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_omni_moe.py:111]: this is going to be a performance downgrade, let's fix this later.

**Review by mingfeima — APPROVED** [review]

**Comment by mingfeima** [comment]
rerun ci to check if the NPU ci fails are related.

**Comment by blzheng** [comment]
> rerun ci to check if the NPU ci fails are related.

The NPU CI failures are unrelated to this PR, as they are also failing on the main branch.

**Comment by mingfeima** [comment]
> > rerun ci to check if the NPU ci fails are related.
> 
> The NPU CI failures are unrelated to this PR, as they are also failing on the main branch.

yes, also failed on main branch: https://github.com/sgl-project/sglang/actions/runs/26378163743/job/77665306423?pr=25644
